dice-where-downloader-lib/src/main/java/technology/dice/dicewhere/downloader/stream/StreamWithMD5Decorator.java

```diff
@@ -25,17 +27,26 @@ public static StreamWithMD5Decorator of(InputStream inputStream) throws NoSuchAl
   }
 
   public MD5Checksum md5() {
-    String hex = (new HexBinaryAdapter()).marshal(this.md5.digest());
-    return MD5Checksum.of(hex);
+    return checksum.orElseGet(
+        () -> {
+          String hex = (new HexBinaryAdapter()).marshal(this.md5.digest());
+          checksum = Optional.of(MD5Checksum.of(hex));
```

Comment:
not a fan of this side effect on `checksum` when calling a "get" method. Is this thread safe?

Reply:
it is not.

Reply:
no, only the one time. we can make it atomic if it's necessary

Reply:
it might get ugly given what you want todo with the `Optional` behaviour